### ISSUE TITLE
(3) Implement the models and train them

### ISSUE THREAD

**Issue by redouane-dziri** [redouane-dziri/deep-music-classification#5]
- [x] Implement a ConvNet with the following specs:
   - Architecture
      - 12 6x6 2D conv filters
      - 2x2 2D average pooling with affine transformation (trainable!)
      - 12 3x3 conv filters
      - 2D global average pooling 
      - 10 dense perceptron with softmax
  - tanh activations

It should leave the input depth as a parameter: most models will have a depth (ie. number of channels) of 1, but i-GLCM will have a depth of 4

- [x] for MFCCM implement a ConvNet with the following specs:

   - 12 10x10 2D conv filters
   - 2x2 2D average pooling with affine transformation (trainable!)
   - 12 3x3 conv filters
   - 2D global average pooling 
   - 10 dense perceptron with softmax

- [x] Make first model able to take depth as parameter for stacked model (i-GLCM)

- [x] Train / debug on sample data

- [x] Train on full data for 400 (!) epochs

- [x] save combined loss curves like in paper

**Comment by arnaudstiegler** [comment]
For having a smooth process, I wrote:

- fetch_data.py: implements a function for fetching the data from the google storage (I will implement the version for local storage as well)
- training_script_template.py: template for fetching the data, putting it into a dataset and training a neural net on it. Should run smoothly

**Comment by redouane-dziri** [comment]
- a notebook that loads the data, trains the model, plots metrics of the training phase and saves the model

**Comment by arnaudstiegler** [comment]
What should we do about validation data during training?

> "In our experiments we use randomly-selected 90 songs from each genre for the training set (in total 900 songs) and the rest for validation (100 songs)."

Should we:
- 1) Train without validation data and use the validation for the final evaluation
- 2) Train with the validation data (not using majority voting), and do the final evaluation with the validation data (using majority voting)
- 3) Train with the validation data with majority voting
- 4) use a data split on the train data and leave validation for the final evaluation

My take on this is that 3) is gonna crush performances and we get not benefit from it compared with 2. 1) could be an issue because we can't control overfitting and 4) could be an issue because we would not be training on the full data.
My take on this is that they did 3) but we should do 2)

**Comment by arnaudstiegler** [comment]
With a simple run on mel maps with angle 0:
- obtained 46.8% accuracy (43% in the paper) on 400 epochs
- could have been better with more epochs (no overfitting yet and no plateau on the loss)
- training process seems to be sensitive to initialization
- decreased the learning rate to 10e-4 (10e-3 was creating oscillations on the training loss)
